// @apple/app-store-server-library/dist/models/AdvancedCommercePeriod.js
"use strict";
// Copyright (c) 2026 Apple Inc. Licensed under MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.AdvancedCommercePeriodValidator = exports.AdvancedCommercePeriod = void 0;
const Validator_1 = require("./Validator");
/**
 * The duration of a single cycle of an auto-renewable subscription.
 *
 * {@link https://developer.apple.com/documentation/advancedcommerceapi/period period}
 */
var AdvancedCommercePeriod;
(function (AdvancedCommercePeriod) {
    AdvancedCommercePeriod["P1W"] = "P1W";
    AdvancedCommercePeriod["P1M"] = "P1M";
    AdvancedCommercePeriod["P2M"] = "P2M";
    AdvancedCommercePeriod["P3M"] = "P3M";
    AdvancedCommercePeriod["P6M"] = "P6M";
    AdvancedCommercePeriod["P1Y"] = "P1Y";
})(AdvancedCommercePeriod || (exports.AdvancedCommercePeriod = AdvancedCommercePeriod = {}));
class AdvancedCommercePeriodValidator extends Validator_1.StringValidator {
}
exports.AdvancedCommercePeriodValidator = AdvancedCommercePeriodValidator;
