// @apple/app-store-server-library/dist/models/AdvancedCommerceRefundReason.js
"use strict";
// Copyright (c) 2026 Apple Inc. Licensed under MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.AdvancedCommerceRefundReasonValidator = exports.AdvancedCommerceRefundReason = void 0;
const Validator_1 = require("./Validator");
/**
 * A reason to request a refund.
 *
 * {@link https://developer.apple.com/documentation/advancedcommerceapi/refundreason refundReason}
 */
var AdvancedCommerceRefundReason;
(function (AdvancedCommerceRefundReason) {
    AdvancedCommerceRefundReason["UNINTENDED_PURCHASE"] = "UNINTENDED_PURCHASE";
    AdvancedCommerceRefundReason["FULFILLMENT_ISSUE"] = "FULFILLMENT_ISSUE";
    AdvancedCommerceRefundReason["UNSATISFIED_WITH_PURCHASE"] = "UNSATISFIED_WITH_PURCHASE";
    AdvancedCommerceRefundReason["LEGAL"] = "LEGAL";
    AdvancedCommerceRefundReason["OTHER"] = "OTHER";
    AdvancedCommerceRefundReason["MODIFY_ITEMS_REFUND"] = "MODIFY_ITEMS_REFUND";
    AdvancedCommerceRefundReason["SIMULATE_REFUND_DECLINE"] = "SIMULATE_REFUND_DECLINE";
})(AdvancedCommerceRefundReason || (exports.AdvancedCommerceRefundReason = AdvancedCommerceRefundReason = {}));
class AdvancedCommerceRefundReasonValidator extends Validator_1.StringValidator {
}
exports.AdvancedCommerceRefundReasonValidator = AdvancedCommerceRefundReasonValidator;
